### ISSUE TITLE
feat: Gradient clipping in standard Trainer

### ISSUE THREAD

**Issue by Samyssmile** [Samyssmile/edux#198]
## Summary

The standard `Trainer` has **no gradient clipping**. Only `DetectionTrainer` implements it. For ImageNet training with warmup or high learning rates, gradient explosions are common and can destabilize training entirely.

### Trainer.java — no clipping
```java
// Training loop (lines 86-109)
for (var batch : trainLoader) {
    var predictions = model.forward(batch.data(), true);
    var lossResult = loss.compute(predictions, batch.labels());
    model.backward(lossResult.gradient());
    optimizer.step(layers);           // ← gradients applied unclipped
    model.clearCaches();
}
```

### DetectionTrainer.java — has clipping (but isolated)
```java
private void clipGradients() {
    if (maxGradNorm <= 0f) return;
    float sumSq = 0f;
    for (var grad : model.allGradients()) {
        float gradNorm = ops.l2Norm(grad);
        sumSq += gradNorm * gradNorm;
    }
    float globalNorm = (float) Math.sqrt(sumSq);
    if (globalNorm > maxGradNorm) {
        float scale = maxGradNorm / globalNorm;
        for (var grad : model.allGradients()) {
            ops.scaleInPlace(grad, scale);
        }
    }
}
```

This logic is duplicated and not available to the standard training pipeline.

## Proposed Solution

Add optional gradient clipping to the `Trainer`:

```java
public final class Trainer {
    private float maxGradNorm = 0f;  // 0 = disabled (default)

    /** Enable global-norm gradient clipping. */
    public Trainer withGradientClipping(float maxGradNorm) {
        this.maxGradNorm = maxGradNorm;
        return this;
    }

    // In training loop, after backward() and before optimizer.step():
    if (maxGradNorm > 0f) {
        clipGradients(model, maxGradNorm);
    }
    optimizer.step(layers);
}
```

### Recommended defaults for common architectures:
| Architecture | maxGradNorm |
|---|---|
| ResNet (ImageNet) | 1.0 |
| Transformer | 1.0 |
| DETR (Detection) | 0.1 |

### Usage
```java
var trainer = new Trainer(model, loss, optimizer)
    .withGradientClipping(1.0f)       // ← simple opt-in
    .withScheduler(scheduler)
    .addCallback(new ProgressCallback(batchCount));
```

## Implementation Notes

- Reuse `ComputeOps.l2Norm()` and `ComputeOps.scaleInPlace()` which already exist
- Extract shared `clipGradients()` logic that `DetectionTrainer` can also delegate to
- Log the global gradient norm in `TrainingContext` for debugging (optional)
- Default `0f` preserves backward compatibility

## Acceptance Criteria

- [ ] `Trainer.withGradientClipping(float maxGradNorm)` enables clipping
- [ ] Clipping happens after `backward()`, before `optimizer.step()`
- [ ] Global L2 norm computed across all parameters
- [ ] Proportional scaling when norm exceeds threshold
- [ ] Default behavior (no clipping) unchanged
- [ ] `DetectionTrainer` refactored to share the same clipping utility
- [ ] Unit test: gradients are scaled when norm exceeds threshold
- [ ] Unit test: gradients unchanged when norm is below threshold

**Comment by Samyssmile** [comment]
Implemented in cf2e17e. Gradient clipping with Norm (L2) and Value (element-wise) strategies, matching PyTorch semantics. Includes GPU-safe zeroInPlace via cudaMemset, SIMD-accelerated CpuOps, and full test coverage.
